docs/source/getting_started.rst

```diff
@@ -103,47 +162,130 @@ Arguments with default values can be annotated as follows:
 
    def greet_all(names: Iterable[str]) -> None:
        for name in names:
-           print('Hello, {}'.format(name))
+           print('Hello ' + name)
 
-For brevity, we often omit the ``typing`` import in code examples, but
+As a final example, suppose want to write a function that can accept *either*
+ints or strings, but no other types. We can express this using the ``Union`` type:
+
+.. code-block:: python
+
+   from typing import Union
+
+   def normalize_id(user_id: Union[int, str]) -> str:
+       if isinstance(user_id, int):
+           return 'user-{}'.format(100000 + user_id)
+       else:
+           return user_id
+
+The ``typing`` module contains many other useful types. You can find a
+quick overview by looking through the :ref:`mypy cheatsheets <overview-cheat-sheets>` 
+and a more detailed overview (including information on how to make your own
+generic types) by looking through the
+:ref:`type system reference <overview-type-system-reference>`.
+
+One final note: when adding types, the convention is to import types 
+using the form ``from typing import Iterable`` (as opposed to doing
+just ``import typing`` or ``import typing as t`` or ``from typing import *``).
+
+For brevity, we often omit these ``typing`` imports in code examples, but
 mypy will give an error if you use definitions such as ``Iterable``
 without first importing them.
 
+Local type inference
+********************
+
+Once you have added type hints to a function (e.g. made it statically typed),
+mypy will automatically type check that function's body. While doing so,
+mypy will try and *infer* as many details as possible. 
+
+We saw an example of this in the ``normalize_id`` function above -- mypy understands
+basic ``isinstance`` checks and so can infer that the ``user_id`` variable was of
+type ``int`` in the if-branch and of type ``str`` in the else-branch.
+
+As another example, consider the following function. Mypy can type-check this function
+without a problem: it will use the available context and deduce that ``output`` must be
+of type ``List[float]`` and that ``num`` must be of type ``float``:
+
+.. code-block:: python
+
+   def nums_below(numbers: Iterable[float], limit: float) -> List[float]:
+       output = []
+       for num in numbers:
+           if num < limit:
+               output.append(num)
+       return output
+
+Mypy will warn you if it is unable to determine the type of some variable.
+For example, suppose we assign an empty dictionary to some global variable:
+
+.. code-block:: python
+
+    my_global_dict = {}  # Error: Need type annotation for 'my_global_dict'
+
+We can teach mypy what type ``my_global_dict`` is meant to have by giving it
+a type hint. For example, if we knew this variable is supposed to be a dict
+of ints to floats, we could annotate it like so:
+
+.. code-block:: python
+
+   # If you are using Python 3.6+, you can use *variable annotations*:
+   my_global_dict: Dict[int, float] = {}
```

Comment:
You can also add a link to PEP 526, like you did above for function annotations.